src/Database/LSMTree/Internal/PageAcc.hs

```diff
@@ -0,0 +1,175 @@
+    , mpOpMap       :: !(P.MutablePrimArray s Word64)   -- ^ operations crumb map
+    , mpBlobRefsMap :: !(P.MutablePrimArray s Word64)   -- ^ blob reference bitmap
+    , mpBlobRefs1   :: !(P.MutablePrimArray s Word64)   -- ^ blob spans 64 bit part (offset)
+    , mpBlobRefs2   :: !(P.MutablePrimArray s Word32)   -- ^ blob spans 32 bit part (size)
+    , mpKeys        :: !() -- TODO: U.MVector s SerializedKey, needs Unbox SerializedKey instance -- ^ array of keys
+    , mpValues      :: !() -- TODO: U.MVector s SerializedValue, needs Unbox SerializedValue instance -- ^ array of values
+    }
+
+keysCountIdx :: Int
+keysCountIdx = 0
+{-# INLINE keysCountIdx #-}
+
+blobRefCountIdx :: Int
+blobRefCountIdx = 1
+{-# INLINE blobRefCountIdx #-}
+
+byteSizeIdx :: Int
+byteSizeIdx = 2
+{-# INLINE byteSizeIdx #-}
+
+maxsizeofEntry :: Int
+maxsizeofEntry = 4052
+{-# INLINE maxsizeofEntry #-}
+
+pageSize :: Int
+pageSize = 4096
+{-# INLINE pageSize #-}
+
+-- | Calculate the total byte size of key, value and optional blobspan.
+--
+-- To fit into single page this has to be at most 4052 (the same as the max key size),
+-- if the entry is larger, the 'pageAccAddElem' would immediately return 'Nothing'.
+--
+-- In other words, if you have a large entry (i.e. Insert with big value), don't use the 'MPageAcc', but construct the single value page directly.
+--
+-- Checking entry size allows us to use 'Word16' arithmetic, we don't need to worry about overflows.
+sizeofEntry :: SerialisedKey -> Entry SerialisedValue BlobSpan -> Int
+sizeofEntry k Delete               = sizeofKey k
+sizeofEntry k (Mupdate v)          = sizeofKey k + sizeofValue v
+sizeofEntry k (Insert v)           = sizeofKey k + sizeofValue v
+sizeofEntry k (InsertWithBlob v _) = sizeofKey k + sizeofValue v + 12
+
+-- | Create new 'MPageAcc'.
+--
+-- The create 'MPageAcc' will be empty.
+newPageAcc :: ST s (MPageAcc s)
+newPageAcc = do
+    mpDir          <- P.newPrimArray 2
+    mpOpMap        <- P.newPrimArray 11 -- 704 / 64
+    mpBlobRefsMap  <- P.newPrimArray 22 -- 704 / 32
+    mpBlobRefs1    <- P.newPrimArray 680
+    mpBlobRefs2    <- P.newPrimArray 680
+    mpKeys         <- return () -- TODO
+    mpValues       <- return () -- TODO
+
+    return MPageAcc {..}
+
+-- | Reuse 'MPageAcc' for constructing new page, the old data will be reset.
+resetPageAcc
+    :: MPageAcc s
+    -> ST s ()
+resetPageAcc MPageAcc {..} = do
+    P.setPrimArray mpDir 0 3 0
+    P.setPrimArray mpOpMap 0 11 0
+    P.setPrimArray mpBlobRefsMap 0 22 0
+    P.setPrimArray mpBlobRefs1 0 680 0
+    P.setPrimArray mpBlobRefs1 0 680 0
+
+-- | Write the first entry to the 'MPageAcc'.
+_addFirstEntry
+    :: MPageAcc s
+    -> SerialisedKey
+    -> Entry SerialisedValue BlobSpan
+    -> ST s ()
+_addFirstEntry MPageAcc {..} k e = assert (sizeofEntry k e <= maxsizeofEntry) $ do
```

Comment:
Hmm, do we want to separate the first from remaining entries? I know the first one can't fail to fit, but is it worth separating it out? Not clear.